[Merged by Bors] - docs: reorganized usage guide and 'implementation.adoc' (#302)
stackabletech/hdfs-operator: docs/modules/ROOT/pages/index.adoc

= Stackable Operator for Apache HDFS

> The Stackable Operator for https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HdfsUserGuide.html[Apache HDFS] is used to set up HFDS in high-availability mode. It depends on the xref:zookeeper:ROOT:index.adoc[] to operate a ZooKeeper cluster to coordinate the active/passive name node.

Reviewer: ```suggestion
The Stackable Operator for https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HdfsUserGuide.html[Apache HDFS] is used to set up HFDS in high-availability mode. It depends on the xref:zookeeper:ROOT:index.adoc[] to operate a ZooKeeper cluster to coordinate the active and standby name nodes.
```

NOTE: This operator only works with images from the https://repo.stackable.tech/#browse/browse:docker:v2%2Fstackable%2Fhadoop[Stackable] repository

== Roles

Three xref:home:concepts:roles-and-role-groups.adoc[roles] of the HDFS cluster are implemented:

> * Data Node - responsible for holding the actual data. *IMPORTANT* Currently the data is kept in `hostPath` volumes of 1GB and no configuration options are exposed to the user. Each data node has its own volume.

Reviewer: I dont understand. The user can configure (recently with https://github.com/stackabletech/hdfs-operator/pull/291 even multiple) volumes for datanodes.
```suggestion
* Data Node - responsible for storing the actual data. *IMPORTANT* Currently the data is kept in `hostPath` volumes of 1GB and no configuration options are exposed to the user. Each data node has its own volume.
```

Author: Maybe that is also outdated then? Do you think the comment should be removed? Maybe we can link to docs that talk about setting up the volumes?

Author: I have removed the statement [de6a8bd](https://github.com/stackabletech/hdfs-operator/pull/302/commits/de6a8bd9d50ff2fd3445a522faf79360f0b16409)

* Journal Node - responsible for keeping track of HDFS blocks and used to perform failovers in case the active name node fails. For details see: https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HDFSHighAvailabilityWithQJM.html
* Name Node - responsible for keeping track of HDFS blocks and providing access to the data.

== Kubernetes objects

The operator creates the following K8S objects per role group defined in the custom resource.

> * ClusterIP - used for intra-cluster communication.

Reviewer: The K8s object is a `Service`
```suggestion
* Service - ClusterIP used for intra-cluster communication.
```

> * ConfigMap - HDFS configuration files like `core-site.xml` and `hdfs-site.xml` are defined here and mounted in the pods.

Reviewer: log4j.properties config as well.

Author: [9957d54](https://github.com/stackabletech/hdfs-operator/pull/302/commits/9957d54d6b97e2ba7aa1f0f571e0ce3dbf5475b6)

> * StatefulSet - where the replica count of each role group is defined. By default, a cluster will have 2 name nodes, 3 journal nodes and 3 data nodes.

Reviewer: Not sure this is true anymore.

Author: [c2c24bc](https://github.com/stackabletech/hdfs-operator/pull/302/commits/c2c24bcc0b18a53e022f0208165446e371e5bb34)

> In addition, a `NodePort` service is created for each pod that exposes all container ports to the outside world (from the perspective of K8S).

Reviewer: ```suggestion
In addition, a `NodePort` service is created for each pod labeled with `hdfs.stackable.tech/pod-service=true` that exposes all container ports to the outside world (from the perspective of K8S).
```
